### DISCUSSION TITLE
Express Sessions not working

### DISCUSSION THREAD

**Discussion by brunoamoroso** [vercel/vercel#11565] [Help]
### Question

I have a checkout flow on vercel which was working fine and now when a customer jumps from one page to another all the req.session gets lost. I noticed that the session ID changes everytime, I know this happens because of the serverless. Although I had contoured this problem using MongoDB as a storage, it came back because it doesn't recognize the session to retrieve the data.

This is my express session settings, when I first have this problem I solved changing sameSite to none, right now I don't have any clue why this is happening again

```js
app.use(session({
  name: "session",
  secret: "secret_example", //not real
  saveUninitialized: false,
  store: store,
  resave: false,
  sameSite: 'none',
  cookie: {
    secure: true,
    maxAge: 360000,
    httpOnly: true,
  }
}));
```

**Comment by chienhsiang-hung** [comment]
Hi @brunoamoroso 

It seems you are experiencing issues with session persistence in your Vercel Express application. The session ID changing with every page jump is indeed indicative of serverless environments where instances may not maintain state between requests. Using MongoDB as a session store should have helped with this issue, but if sessions are not being recognized, it could be related to a configuration or deployment issue.

There are suggestions that involve checking the trust proxy setting in Express when behind proxies, which is common in serverless deployments like Vercel. This setting is crucial for secure cookie handling. https://stackoverflow.com/questions/65767024/express-session-not-working-in-production-deployment Additionally, ensuring that your code is optimized and using caching can help with serverless function timeout errors. https://devcodef1.com/news/1088606/vercel-serverless-function-timeout-errors

Here are some steps you could consider:
1. Verify the trust proxy setting in your Express application.
2. Review your MongoDB session store configuration to ensure it's correctly storing and retrieving session data.
3. Check for any recent changes in Vercel's deployment environment that might affect session handling.
4. Consult with colleagues who have worked on similar issues or with session management in serverless environments.

If you have already tried these steps or if they do not resolve the issue, it may be beneficial to reach out to someone with specific expertise in Vercel deployments or Express session management.

**Reply by brunoamoroso** [reply]
Hi @chienhsiang-hung thanks for your reply. I've tried everything there. My cookies were first set based on that question that how I solved at the first time this problem. And now I'm facing this again.

I've added the trust proxy that I didn't need before just to try, but got no lucky on that. I've set the maxDuration of the function timeout to 60s and now I don't get this error anymore. Although my problem with the session data that I need keep happening. I really don't what else I could do. 

```js
export const maxDuration = 60;

app.set('trust proxy', 1);

app.use(session({
  name: "session",
  secret: "justus_secret",
  saveUninitialized: false,
  resave: false,
  proxy: true,
  store: store,
  sameSite: 'none',
  cookie: {
    secure: true,
    maxAge: 360000,
  }
}));
```

**Reply by vsreddyh** [reply]
Hi @brunoamoroso i am also facing the same problem. If you have found a fix please share.

**Reply by brunoamoroso** [reply]
Hi @vsreddyh unfortunately I didn't find any solutions for this problem and afterwards I decided to cancel the project that I was getting this error.
